api/server/errorhandler.go

```diff
@@ -17,18 +21,80 @@ func makeErrorHandler(err error) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		statusCode := httpstatus.FromError(err)
 		vars := mux.Vars(r)
-		if apiVersionSupportsJSONErrors(vars["version"]) {
-			response := &types.ErrorResponse{
-				Message: err.Error(),
-			}
-			_ = httputils.WriteJSON(w, statusCode, response)
-		} else {
+		version := vars["version"]
+
+		const firstAPIVersionWithJSONErrors = "1.24"
+		if version != "" && versions.LessThan(version, firstAPIVersionWithJSONErrors) {
 			http.Error(w, status.Convert(err).Message(), statusCode)
+			return
+		}
+
+		response := marshalErrorResponse(err)
+
+		const firstAPIVersionWithJSONMultiErrors = "1.44"
+		if version != "" && versions.LessThan(version, firstAPIVersionWithJSONMultiErrors) {
+			response.Errors = nil
 		}
+
+		_ = httputils.WriteJSON(w, statusCode, response)
 	}
 }
 
-func apiVersionSupportsJSONErrors(version string) bool {
-	const firstAPIVersionWithJSONErrors = "1.23"
-	return version == "" || versions.GreaterThan(version, firstAPIVersionWithJSONErrors)
+type joiningError interface {
+	Unwrap() []error
+	Error() string
+}
+
+type wrappingError interface {
+	Unwrap() error
+	Error() string
+}
+
+// marshalErrorResponse returns a tree of [*types.ErrorResponse] representing the argument err.
+func marshalErrorResponse(err error) *types.ErrorResponse {
+	if errdefs.IsHTTPOnlyError(err) {
+		if wrappedErr := errors.Unwrap(err); wrappedErr != nil {
+			err = wrappedErr
+		}
+	}
+
+	if err, ok := err.(joiningError); ok {
+		return marshalJoinedErrors(err)
+	}
+
+	if err, ok := err.(wrappingError); ok {
+		return &types.ErrorResponse{
+			Message: err.Error(),
+			Errors:  []*types.ErrorResponse{marshalErrorResponse(err.Unwrap())},
+		}
+	}
+
+	return &types.ErrorResponse{
+		Message: err.Error(),
+	}
+}
+
+func marshalJoinedErrors(err joiningError) *types.ErrorResponse {
+	var subErrors []*types.ErrorResponse
+	for _, subErr := range err.Unwrap() {
+		if subErr != nil {
+			subErrors = append(subErrors, marshalErrorResponse(subErr))
+		}
+	}
+
+	errMsg := err.Error()
+	if len(subErrors) > 1 {
+		var stringErrs []string
+		for _, subErr := range subErrors {
+			stringErrs = append(stringErrs, strings.Replace(subErr.Message, "\n", "\n\t", -1))
+		}
+		errMsg = fmt.Sprintf("%d errors occurred:\n\t* %s", len(subErrors), strings.Join(stringErrs, "\n\t* "))
```

Comment:
You can't unconditionally clobber the error message as the error message could carry information not contained within the wrapped errors. Case in point: `fmt.Errorf("foo: %w, %w", a, b)` returns an error which satisfies `joiningError` but whose `Error()` is the formatted string. It would be safe to replace `errMsg` if and only if `err.Error() == errors.Join(err.Unwrap()...).Error()`.